main/actions/src/main/scala/sbt/Compiler.scala

```diff
@@ -106,14 +106,14 @@ object Compiler {
         (Analysis.empty(nameHashing = incOptions.nameHashing), None)
       }
     }
-    compile(InputsWithPrevious(in, PreviousAnalysis(previousAnalysis, previousSetup)), log, reporter).analysis
+    compile(InputsWithPrevious(in, PreviousAnalysis(previousAnalysis, previousSetup)), log, reporter, _.compile).analysis
   }
-  def compile(in: InputsWithPrevious, log: Logger): CompileResult =
+  def compile(in: InputsWithPrevious, log: Logger, compile_ : MixedAnalyzingCompiler => (Set[File], DependencyChanges, xsbti.AnalysisCallback) => Unit = _.compile): CompileResult =
```

Comment:
As you may have noticed, there's a TON of overridden methods in here.   This is because we can't break binary compatibilty, which is why the API is a bit clunky (and why we have too many layers and can't remove/consolidate them).

Do you have time to do an in-person chat about this feature? I'd like to see more what you're trying to do, and look into an alternative design (as an iterim).

Note: in 1.0 we hope to get 99% of this API int' the `actoins/Compile.scala` so you can have your hook way up there.   Additionally IC vs. IncrementalCompiler distinction should disappear.